### PR TITLE
Laboratory work #1, Anastasia Logunova - 22FPL1

### PR DESCRIPTION
**Pull request by logunova-anastasia**

### CODE REVIEW

File: lab_1_classify_by_unigrams/main.py [hunk 1]
```diff
@@ -11,6 +11,16 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:14]: let use early return : https://medium.com/swlh/return-early-pattern-3d18a41bba8

File: lab_1_classify_by_unigrams/main.py [hunk 2]
```diff
@@ -11,6 +11,16 @@ def tokenize(text: str) -> list[str] | None:
...
+        punctuation = '''!()-[]{};:'",<>./?@#$%^&*_~0123456789\\ '''
+        tokenized = []
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:16]: Take a look at list comprehension

File: lab_1_classify_by_unigrams/main.py [hunk 3]
```diff
@@ -19,6 +29,16 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if isinstance(tokens, list):
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:32]: the same

File: lab_1_classify_by_unigrams/main.py [hunk 4]
```diff
@@ -19,6 +29,16 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
...
+        frequency = {}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:33]: and dict comprehension here

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]
Great start. Continue, please!

File: lab_1_classify_by_unigrams/main.py (hunk 1 above)
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 2 above)
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 3 above)
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 4 above)
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py [hunk 5]
```diff
@@ -28,6 +40,11 @@ def create_language_profile(language: str, text: str) -> dict[str, str | dict[st
     :param text: a text
     :return: a dictionary with two keys – name, freq
     """
+    if not isinstance(language, str) or not isinstance(text, str):
+        return None
+    tokens = tokenize(text)
+    profile = {'name': language, 'freq': calculate_frequencies(tokens)}
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:46]: again - return without creating a variable

File: lab_1_classify_by_unigrams/main.py [hunk 6]
```diff
@@ -11,6 +11,11 @@ def tokenize(text: str) -> list[str] | None:
     :param text: a text
     :return: a list of lower-cased tokens without punctuation
     """
+    if not isinstance(text, str):
+        return None
+    punctuation = '''!()-[]{};:'",<>./?@#$%^&*_~0123456789\\ '''
+    tokenized = [el for el in text.lower() if el not in punctuation]
+    return tokenized
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py [hunk 7]
```diff
@@ -19,6 +24,13 @@ def calculate_frequencies(tokens: list[str] | None) -> dict[str, float] | None:
     :param tokens: a list of tokens
     :return: a dictionary with frequencies
     """
+    if not isinstance(tokens, list):
+        return None
+    for token in tokens:
+        if not isinstance(token, str):
+            return None
+    frequency = {token: tokens.count(token) / len(tokens) for token in tokens}
+    return frequency
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py:33]: Done

File: lab_1_classify_by_unigrams/main.py (hunk 5 above)
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py [hunk 8]
```diff
@@ -37,6 +54,17 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
...
+    if not isinstance(predicted, list) or not isinstance(actual, list) \
+            or len(predicted) != len(actual):
+        return None
+    zipped_values = zip(predicted, actual)
+    zipped_list = list(zipped_values)
+    summation = 0
+    for i in zipped_list:
+        summation += (i[1] - i[0]) ** 2
+    number = len(predicted)
+    mse = summation / number
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py [hunk 9]
```diff
@@ -37,6 +55,15 @@ def calculate_mse(predicted: list, actual: list) -> float | None:
     :param actual: a list of actual values
     :return: the score
     """
+    if not isinstance(predicted, list) or not isinstance(actual, list) \
+            or len(predicted) != len(actual):
+        return None
+    zipped_values = zip(predicted, actual)
+    zipped_list = list(zipped_values)
```
[the remark this answers is not in the document]
Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: Done

File: lab_1_classify_by_unigrams/main.py [hunk 10]
```diff
@@ -80,6 +136,15 @@ def preprocess_profile(profile: dict) -> dict[str, str | dict] | None:
     :return: a dict with a lower-cased loaded profile
     with relative frequencies without unnecessary ngrams
     """
+    if not isinstance(profile, dict) or not ('freq' or 'name' or 'n_words') in profile:
+        return None
+    edited_profile = {'name': profile['name'], 'freq': {}}
+    for token in profile['freq']:
+        if token.lower() in edited_profile['freq']:
+            edited_profile['freq'][token.lower()] += profile['freq'][token]/profile['n_words'][0]
```
Comment (artyomtugaryov) [inline lab_1_classify_by_unigrams/main.py:144]: save `token.lower()`, `profile['freq'][token]`, `profile['n_words'][0]` to variables

Comment (logunova-anastasia) [reply lab_1_classify_by_unigrams/main.py]: done!

**Pull request closed without merging**